### PR TITLE
Refactor makeTerraformInputs and friends

### PR DESCRIPTION
**Pull request by VenelinMartinov**
This PR refactors the various functions related to converting pulumi inputs to TF inputs.

It also exposes a new option when making inputs: `EnableMaxItemsOneDefaults` which will allow us to fix https://github.com/pulumi/pulumi-aws/issues/3427 once we can properly test the changes.

Opted not to land the behavioural changes for `EnableMaxItemsOneDefaults` yet because of the multiple rounds of varied downstream regressions - seems like we need to add a framework for testing against the TF behaviour in order to land this.

### CODE REVIEW

**Comment by VenelinMartinov** [comment]
The downstream checks revealed at least 2 new issues, which weren't known before, perhaps due to https://github.com/pulumi/pulumi-terraform-bridge/issues/1730

https://github.com/pulumi/pulumi-vault/actions/runs/8159199780/job/22303439815?pr=429
https://github.com/pulumi/pulumi-azure/actions/runs/8158876277/job/22303367852?pr=1827

File: pkg/tfbridge/schema_test.go [hunk 1]
```diff
@@ -411,6 +411,135 @@ func TestMakeTerraformInputMixedMaxItemsOne(t *testing.T) {
 	}
 }
 
+func TestMaxItemsOneEmptyOldState(t *testing.T) {
```
Comment (t0yv0) [inline pkg/tfbridge/schema_test.go:414]: Can we name this test in a way that tells the reader what it's testing, perhaps assisted by a comment? Something like this:

Test that makeTerraformInputs, makeTerraformInputsWithDefaults work well with MaxItems=1 properties.

- missing MaxItems=1 properties should present to TF as empty collections when applying the actual changes (makeTerraformInputsWithDefaults)
- missing MaxItems=1 properties should present to TF as missing when running validators (makeTerraformInputs)

Nit: it could also benefit from a table-driven test pattern to eliminate a little bit of code repetition.

File: pkg/tfbridge/provider_test.go [hunk 2]
```diff
@@ -3259,3 +3264,223 @@ func TestSchemaFuncsNotCalledDuringRuntime(t *testing.T) {
...
+		resources: map[tokens.Type]Resource{
+			"Res": {
+				TF:     shimv2.NewResource(p.ResourcesMap["res"]),
+				TFName: "res",
+				Schema: &ResourceInfo{},
+			},
+		},
+	}
+
+	t.Run("No error when plural specified", func(t *testing.T) {
```
Comment (t0yv0) [inline pkg/tfbridge/provider_test.go:3455]: What does it mean "plural specified"?

**Review by t0yv0 — APPROVED** [review]
This is great, thank you very much. Conceptually this makes a lot of sense to me - the missing MaxItems=1 properties should be presenting as empty collections to the Terraform provider but not to the provider's validators. We have a slightly convoluted interaction with these validators already so the fix has to accommodate. This is really driving toward the correct behavior though!  Let's do a pass to try to increase the longevity of the tests so we can still keep these tests around and understand what they're testing as the implementation itself evolves, this is an important property - left some style nits and suggestions, otherwise LGTM.

**Review by iwahbe — APPROVED** [review]
I agree with all of t0yv0's comments.

File: pkg/tfbridge/provider_test.go (hunk 2 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/provider_test.go:3457]: mb, this was copy pasted from a test I ended up deleting.

File: pkg/tfbridge/provider.go [hunk 3]
```diff
@@ -517,7 +517,7 @@ func buildTerraformConfig(ctx context.Context, p *Provider, vars resource.Proper
 		}
 	}
 
-	inputs, _, err := MakeTerraformInputs(ctx, nil, tfVars, nil, tfVars, p.config, p.info.Config)
+	inputs, _, err := makeTerraformInputsWithoutTFDefaults(ctx, nil, tfVars, nil, tfVars, p.config, p.info.Config)
```
Comment (VenelinMartinov) [inline pkg/tfbridge/provider.go:520]: Note I've now changed `CheckConfig` to not get the TF defaults - this addresses half of https://github.com/pulumi/pulumi-terraform-bridge/issues/1603 and gets the right behaviour WRT empty MaxItemsOne config

Comment (VenelinMartinov) [reply pkg/tfbridge/provider.go]: I've had to revert this - it doesn't work. Instead I added a `makeTerraformInputsWithoutMaxItemsOneDefaults` variant which is used here - it applies TF defaults but not the maxItemsOne empty arrays.

**Comment by VenelinMartinov** [comment]
Very exciting, another round of failures. I should not have merged master here...

Panic in azure: https://github.com/pulumi/pulumi-azure/actions/runs/8177430025/job/22360306506?pr=1836

some unexpected diffs:
https://github.com/pulumi/pulumi-auth0/actions/runs/8177040722/job/22358337565?pr=451
https://github.com/pulumi/pulumi-azure/actions/runs/8177430025/job/22360306506?pr=1836

Attribute must be a single value not a list:
https://github.com/pulumi/pulumi-aws/actions/runs/8177823475/job/22362420833?pr=3583

And some failing GRPC tests in AWS, which look fine.

**Comment by t0yv0** [comment]
Keep up the good work! It's excellent that the downstream checks are shielding our users from regressions. LMK if you want help.

**Comment by VenelinMartinov** [comment]
The latest round of downstream tests passed aside from RKE and AWS and the AWS failure looks like a flake. This should be good to go now.

File: pkg/tfbridge/schema.go [hunk 4]
```diff
@@ -1228,6 +1258,7 @@ func MakeTerraformConfig(ctx context.Context, p *Provider, m resource.PropertyMa
 	return MakeTerraformConfigFromInputs(ctx, p.tf, inputs), cctx.Assets, nil
 }
 
+// TODO: this is now unused internally, can we delete it?
```
Comment (VenelinMartinov) [inline pkg/tfbridge/schema.go:1261]: This would technically be a breaking change, are we fine to delete this or should we keep it?

File: pkg/tfbridge/provider.go [hunk 5]
```diff
@@ -1013,15 +1009,20 @@ func (p *Provider) Create(ctx context.Context, req *pulumirpc.CreateRequest) (*p
...
-		fmt.Sprintf("%s.news", label))
+	inputs, assets, err := makeTerraformInputsNoDefaultsWithMaxItemsOneDefaults(
+		ctx, nil, nil, nil, props, res.TF.Schema(), res.Schema.Fields,
+	)
 	if err != nil {
-		return nil, errors.Wrapf(err, "preparing %s's new property state", urn)
+		return nil, errors.Wrapf(err, "preparing %s's new property inputs", urn)
 	}
-
+	config := MakeTerraformConfigFromInputs(ctx, p.tf, inputs)
```
Comment (VenelinMartinov) [inline pkg/tfbridge/provider.go:1025]: I've inlined UnmarshalTerraformConfig here since this was the only place it was used and it limited customization here. `makeTerraformInputsNoDefaultsWithMaxItemsOneDefaults` is a new `makeTerraformInputs` variant used only here. This is meant to be the only logic change to this method.

File: pkg/tests/regress_aws_1423_test.go [hunk 6]
```diff
@@ -27,6 +27,7 @@ import (
 )
 
 func TestRegressAws1423(t *testing.T) {
+	t.Skip("Refresh is dirty on this resource see https://github.com/pulumi/pulumi-aws/issues/3361.")
```
Comment (iwahbe) [inline pkg/tests/regress_aws_1423_test.go:30]: This is fine. We are going to remove this test in the near future anyway. ([link](https://github.com/pulumi/pulumi-terraform-bridge/pull/1745#issuecomment-1986148079))

File: pkg/tfbridge/provider.go [hunk 7]
```diff
@@ -1013,15 +1009,20 @@ func (p *Provider) Create(ctx context.Context, req *pulumirpc.CreateRequest) (*p
...
+		plugin.MarshalOptions{Label: label, KeepUnknowns: true, SkipNulls: true})
+	if err != nil {
+		return nil, errors.Wrapf(err, "unmarshaling %s's new property state", urn)
+	}
 	// To get Terraform to create a new resource, the ID must be blank and existing state must be empty (since the
 	// resource does not exist yet), and the diff object should have no old state and all of the new state.
-	config, assets, err := UnmarshalTerraformConfig(ctx,
-		p, req.GetProperties(), res.TF.Schema(), res.Schema.Fields,
-		fmt.Sprintf("%s.news", label))
+	inputs, assets, err := makeTerraformInputsNoDefaultsWithMaxItemsOneDefaults(
```
Comment (iwahbe) [inline pkg/tfbridge/provider.go:1019]: It feels like we have reached the point where we should replace `makeTerraformInputsHelper` with 
```go
makeTerraformInputsWithOptions(..., makeTerraformInputOptions{
	Defaults: False(),
	MaxItemsOneDefaults: True(),
})
```

`makeTerraformInputsNoDefaultsWithMaxItemsOneDefaults` is pretty long and hard to read. I first read it as `makeTerraformInputsNoDefaultsWithMaxItemsOneNoDefaults`.

We would then fold `makeTerraformInputsWithoutTFDefaults` into the same scheme.

File: pkg/tfbridge/provider_test.go [hunk 8]
```diff
@@ -3259,3 +3260,420 @@ func TestSchemaFuncsNotCalledDuringRuntime(t *testing.T) {
...
+func TestProviderCheckConfigRequiredDefaultEnvConfig(t *testing.T) {
+	t.Setenv("REQUIRED_CONFIG", "required")
+	// Note that this config should be invalid.
+	//
+	// From the Required docs: Required cannot be used with Computed Default, DefaultFunc
+	// attributes in a Provider schema
+	//
+	// From the DefaultFunc docs: For legacy reasons, DefaultFunc can be used with Required
+	//
+	// This is needed right now since some providers (e.g. Azure) depend on this.
```
Comment (iwahbe) [inline pkg/tfbridge/provider_test.go:3581]: 🎉 This is excellent documentation! 🎉

File: pkg/tfbridge/schema.go (hunk 4 above)
Comment (iwahbe) [reply pkg/tfbridge/schema.go:1252]: I'm ok removing it. We've made breaking changes to this function recently (introducing `ctx`) and any consumers can easily inline the function.

This is one of those functions that really shouldn't have been public to begin with.

File: pkg/tfbridge/schema_test.go [hunk 9]
```diff
@@ -42,30 +42,22 @@ import (
 	shimv2 "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/sdk-v2"
 )
 
-func makeTerraformInputs(olds, news resource.PropertyMap,
```
[the remark this answers is not in the document]
Comment (iwahbe) [reply pkg/tfbridge/schema_test.go:45]: Is there a reason why we don't just use the production methods directly, removing `tMakeTerraformInputsWithoutAnyDefaults` and `makeTerraformInputs`?

File: pkg/tfbridge/provider.go (hunk 7 above)
Comment (iwahbe) [reply pkg/tfbridge/provider.go:1019]: If we want to name these functions, let's name them something that describes where they should be used, not just the flags that we have set. For example, we can express this excellent comment (borrowed from a test case) like this:

```go
// missing MaxItems=1 properties should present to TF as empty collections when creating
// a resource (makeTerraformInputsNoDefaultsWithMaxItemsOneDefaults)
```

```go
// makeTerraformInputsForResourceCreate translates Pulumi property values 
// to the form that TF expects for a resource creation.
func makeTerraformInputsForResourceCreate(...) ... {
	return makeTerraformInputsWithOptions(..., makeTerraformInputOptions{
	Defaults: False(),
	MaxItemsOneDefaults: True(),
}
```

File: pkg/tfbridge/schema_test.go [hunk 10]
```diff
@@ -411,6 +403,125 @@ func TestMakeTerraformInputMixedMaxItemsOne(t *testing.T) {
 	}
 }
 
+// Test that makeTerraformInputs variants work well with MaxItems=1 properties.
+// missing MaxItems=1 properties should present to TF as missing in most cases (MakeTerraformInputs)
+// missing MaxItems=1 properties should present to TF as missing
+// when running validators (makeTerraformInputsWithoutTFDefaults)
+// missing MaxItems=1 properties should present to TF as empty collections when creating
+// a resource (makeTerraformInputsNoDefaultsWithMaxItemsOneDefaults)
+func TestMakeTerraformInputsWithMaxItemsOne(t *testing.T) {
```
Comment (iwahbe) [inline pkg/tfbridge/schema_test.go:412]: Awesome!

**Review by iwahbe — COMMENTED** [review]
The PR looks great. Thanks for running this all down. 

Have you tested that AWS passes tests with https://github.com/pulumi/pulumi-aws/issues/3427 removed?

There are a couple of questions I have about style, but the functionality looks great.

File: pkg/tfbridge/schema_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (VenelinMartinov) [reply pkg/tfbridge/schema_test.go:45]: The test wrappers reduce the boilerplate in the tests a bit.

File: pkg/tfbridge/provider.go (hunk 7 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/provider.go]: I've refactored the makTerraformInputs* as suggested, LMK if it now makes more sense.

File: pkg/tests/regress_940_test.go
Comment (VenelinMartinov) [inline pkg/tests/regress_940_test.go:1]: This was moved to `schema_test.go` since it uses `schema`-internal functions

**Review by iwahbe — APPROVED** [review]

File: pkg/tfbridge/provider.go (hunk 7 above)
Comment (iwahbe) [reply pkg/tfbridge/provider.go]: Looks good to me.

File: pkg/tfbridge/provider.go [hunk 11]
```diff
@@ -780,18 +780,17 @@ func (p *Provider) Check(ctx context.Context, req *pulumirpc.CheckRequest) (*pul
 	rescfg := MakeTerraformConfigFromInputs(ctx, p.tf, inputs)
 	warns, errs := p.tf.ValidateResource(ctx, tfname, rescfg)
 	for _, warn := range warns {
-		if err = p.host.Log(ctx, diag.Warning, urn, fmt.Sprintf("%v verification warning: %v", urn, warn)); err != nil {
-			return nil, err
-		}
+		GetLogger(ctx).Warn(fmt.Sprintf("%v verification warning: %v", urn, warn))
```
Comment (t0yv0) [inline pkg/tfbridge/provider.go:783]: Is this line intentionally changing?

File: pkg/tfbridge/provider.go [hunk 12]
```diff
@@ -1013,15 +1012,21 @@ func (p *Provider) Create(ctx context.Context, req *pulumirpc.CreateRequest) (*p
...
+		return nil, errors.Wrapf(err, "unmarshaling %s's new property state", urn)
+	}
 	// To get Terraform to create a new resource, the ID must be blank and existing state must be empty (since the
 	// resource does not exist yet), and the diff object should have no old state and all of the new state.
-	config, assets, err := UnmarshalTerraformConfig(ctx,
-		p, req.GetProperties(), res.TF.Schema(), res.Schema.Fields,
-		fmt.Sprintf("%s.news", label))
+	inputs, assets, err := makeTerraformInputsWithOptions(
+		ctx, nil, nil, nil, props, res.TF.Schema(), res.Schema.Fields,
+		makeTerraformInputsOptions{Defaults: False(), TFDefaults: False(), MaxItemsOneDefaults: True()},
```
Comment (t0yv0) [inline pkg/tfbridge/provider.go:1024]: This is a very large change here that seems unrelated to the goals of the PR.  UnmarshalTerraformConfig used to apply defaults and it doesn't do that anymore. Why is it needed? Why do we believe it is safe?

Comment (t0yv0) [inline pkg/tfbridge/provider.go:1024]: Having only Create opt into `MaxItemsOneDefaults: True()` is very suspicious. Why is it needed? Would it not guarantee non-empty updates where the created resource is compared via Update with itself, and Update does not perform the MaxItemsOneDefaults transformation, causing diffs?

File: pkg/tests/regress_940_test.go [hunk 13]
```diff
@@ -1,81 +0,0 @@
...
-	"testing"
-
-	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
-	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge"
-	shimv2 "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/sdk-v2"
-	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
-	"github.com/stretchr/testify/assert"
-)
-
-func TestRegress940(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (VenelinMartinov) [reply pkg/tests/regress_940_test.go:28]: This is just a move - I've moved the test to schema_test.go since I made the function it uses private.

I had a comment on the PR about it somehow got lost along the way.

**Comment by t0yv0** [comment]
I have difficulty understanding/verifying the latest version of this PR. Let's perhaps do a session tomorrow to get on the same page. Thank you.

File: pkg/tfbridge/provider.go (hunk 11 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/provider.go:783]: Yeah, host is nil in tests which causes this to panic. `GetLogger` seems to handle that but it's not obvious it logs to the same place. I'll revert to manually handling this for now, we can ship this later if this is appropriate.
cc @iwahbe

File: pkg/tfbridge/provider.go (hunk 12 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/provider.go]: I don't believe `UnmarshalTerraformConfig` applied defaults but I agree it's not obvious:

It called `MakeTerraformConfig`:
https://github.com/pulumi/pulumi-terraform-bridge/blob/54802d5b003cd46ff6163a3367f5b8815217eacf/pkg/tfbridge/schema.go#L1241

which does not set the Defaults in the conversion context:

https://github.com/pulumi/pulumi-terraform-bridge/blob/54802d5b003cd46ff6163a3367f5b8815217eacf/pkg/tfbridge/schema.go#L1219-L1223

so they should not get applied

**Comment by VenelinMartinov** [comment]
After discussing with @t0yv0 I'll add a test to ensure we don't generate spurious update for Create -> Update for the same resource set.

I'll also open a follow-up item on investigating TF behaviour around the other methods, most interestingly Update and figure out how TF handles missing MaxItemsOne properties in Update - it might need the same handling as Create.

**Review by t0yv0 — APPROVED** [review]
Thank for jumping on a call. This is very fascinating, we recovered this call graph:

```
UnmarshalTerraformConfig did not apply defaults

 UnmarshalTerraformConfig calls
 MakeTerraformConfig creates
 conversionContext{} without apply defaults
 calls cctx.makeTerraformInputs

 The one that does apply defaults is called

 buildTerraformConfig calls
 MakeTerraformInputs

 Check calls it GAH.
 CheckConfig calls buildTerraformConfig

 Invoke calls it.
```

So this PR does not affect not applying defaults in Create. 

AS we discussed I have a hunch that Update should match Create in how it behaves, for symmetry and also for matching TF behavior on update, but this needs to be tested out.

**Comment by VenelinMartinov** [comment]
Just tested the original issue with this PR and without the patches. @t0yv0 was absolutely right about `Update` - it looks like it also needs the same treatment as `Create` - the original panic no longer happens on `Create` but now triggers on `Update`.

I'll update the PR and run another round of downstream checks

**Comment by t0yv0** [comment]
Basically I trust that this can be merged when passes checks but could you please rewrite the commit description to reflect the latest version so we have a good message in the commit log (git squash makes your PR descriptions into commit messages - can do that with git CLI obviously just so we get the desired end-result). TYVM.

**Comment by VenelinMartinov** [comment]
Downstream tests yielded a panic in azure: https://github.com/pulumi/pulumi-azure/actions/runs/8271477043/job/22632451751?pr=1864

As well as what looks like unrelated errors in keycloak: https://github.com/pulumi/pulumi-keycloak/actions/runs/8271706418/attempts/1?pr=425

Looking into the panic.

**Comment by t0yv0** [comment]
I'm so sorry. 😢

**Comment by VenelinMartinov** [comment]
The azure panic is somewhat interesting - it happens here: https://github.com/hashicorp/terraform-provider-azurerm/blob/61ac20e312db3e460ccb2324fdc04bc9eb2e503e/internal/services/eventhub/eventhub_namespace_resource.go#L510

`ruleSets` is now an empty array so `rulesets` becomes a nil struct, so `*rulesets.Properties.PublicNetworkAccess` is a nil value dereference.

This is not great but I tried to see if this was an error before. Seems like previously we were unable to actually hit this because of https://github.com/pulumi/pulumi-azure/issues/1881.

So my suspicion is that this was wrong before it is a bit less wrong now but panics.

I also suspect that the we should be doing `MaxItemsOneDefaults` when we `makeTerraformState` in `Update` since otherwise there'd be a diff between the state and news

**Comment by VenelinMartinov** [comment]
From https://github.com/pulumi/terraform-plugin-sdk/blob/7ac578ce47fc07e0888beee6d4ab9db09369274f/helper/schema/schema.go#L78-L96:


```
If Elem is *schema.Resource then setting ConfigMode to SchemaConfigModeAttr will force it to be represented in configuration as an attribute, which means that the Computed flag can be used to provide default elements when the argument isn't set at all, while still allowing the user to force zero elements by explicitly assigning an empty list.
```

I have a feeling this is used to set a default for the rulesets and that's the reason TF does not error out here. The resource https://github.com/hashicorp/terraform-provider-azurerm/blob/61ac20e312db3e460ccb2324fdc04bc9eb2e503e/internal/services/eventhub/eventhub_namespace_resource.go#L128
has `ConfigMode: pluginsdk.SchemaConfigModeAttr`.


Will do a more in-depth comparison of the TF behaviour here with what the bridge does next week.

**Comment by VenelinMartinov** [comment]
Terraform docs on `SchemaConfigModeAttr`:
https://developer.hashicorp.com/terraform/plugin/sdkv2/guides/terraform-0.12-compatibility#computed-resource-attributes

```
Only activate this mode if your provider has existing functionality that is depending on the ability to distinguish unset vs. explicitly empty for a nested resource collection.
```
which sounds very very relevant.

**Comment by VenelinMartinov** [comment]
Old description for posterity: This reverts commit 624167331a8fd049f2bf589865f7dcea19dd535e and gets https://github.com/pulumi/pulumi-terraform-bridge/pull/1688 back in.

This should allow us to address https://github.com/pulumi/pulumi-aws/issues/3427 and unblock further upstream updates there.

The issue was that TF seems to pass empty arrays for not-specified MaxItemsOne properties, while we were passing nil, resulting in differences in behaviour between the bridge and TF.

The original PR went a bit too far and made the change for `Check` too, which caused some issues: https://github.com/pulumi/pulumi-aws/issues/3427#issuecomment-1976851184.

I've reused the mechanism added https://github.com/pulumi/pulumi-terraform-bridge/pull/1583 and only apply the empty arrays for MaxItemsOne properties in makeTerraformInputs when the `ApplyMaxItemsOneDefaults` flag is set. This makes sure we don't mess with the parameters when doing `Check` but pass the correct parameters to other methods.

This PR introduces a new `makeTerraformInputs` variant: `makeTerraformInputsWithoutMaxItemsOneDefaults` - this is only used for `CheckConfig`, since `CheckConfig` depends on the null MaxItemsOne properties but needs the TF defaults.

**Comment by VenelinMartinov** [comment]
I've opened https://github.com/pulumi/pulumi-terraform-bridge/issues/1767 for the follow up work.

File: pkg/tfbridge/provider.go [hunk 14]
```diff
@@ -780,18 +780,24 @@ func (p *Provider) Check(ctx context.Context, req *pulumirpc.CheckRequest) (*pul
 	rescfg := MakeTerraformConfigFromInputs(ctx, p.tf, inputs)
 	warns, errs := p.tf.ValidateResource(ctx, tfname, rescfg)
 	for _, warn := range warns {
-		if err = p.host.Log(ctx, diag.Warning, urn, fmt.Sprintf("%v verification warning: %v", urn, warn)); err != nil {
-			return nil, err
+		warning := fmt.Sprintf("%v verification warning: %v", urn, warn)
+		if p.host == nil {
```
Comment (t0yv0) [inline pkg/tfbridge/provider.go:784]: Probably a code-level TODO here that this temporarily required for tests, before we can inject a mock host in the tests.. Ideally we'd not make the prod code more complex for tests.

File: pkg/tfbridge/provider.go [hunk 15]
```diff
@@ -1013,15 +1019,20 @@ func (p *Provider) Create(ctx context.Context, req *pulumirpc.CreateRequest) (*p
...
-	config, assets, err := UnmarshalTerraformConfig(ctx,
-		p, req.GetProperties(), res.TF.Schema(), res.Schema.Fields,
-		fmt.Sprintf("%s.news", label))
+	config, assets, err := makeTerraformConfigWithOpts(
+		ctx, p, props, res.TF.Schema(), res.Schema.Fields,
+		makeTerraformConfigOpts{},
+	)
 	if err != nil {
-		return nil, errors.Wrapf(err, "preparing %s's new property state", urn)
+		return nil, errors.Wrapf(err, "preparing %s's new property inputs", urn)
```
Comment (t0yv0) [inline pkg/tfbridge/provider.go:1034]: This is good, this was a typo before looks like.

File: pkg/tfbridge/schema.go [hunk 16]
```diff
@@ -655,6 +650,18 @@ func (ctx *conversionContext) makeObjectTerraformInputs(
 		return nil, err
 	}
 
+	if tfs != nil && ctx.ApplyMaxItemsOneDefaults {
```
Comment (t0yv0) [inline pkg/tfbridge/schema.go:653]: As discussed this code is not enabled yet as nothing sets ctx.ApplyMaxItemsOneDefaults ; instead we're going to try to land cross-tests first to increase coverage in this area significantly before iterating further through downstream tests.

**Comment by t0yv0** [comment]
I was a bit uncomfortable at first but approving after our chat and another round of looking at it in detail. Behavior is not changing, some new code is added but it's flagged off, and new tests are added which is excellent.

A bit of a nit but I'd prefer in the future not to remove public members `MakeTerraformState` cold-turkey but mark them deprecated or call out they're for internal use and subject to change in a comment. It does not take a lot more time and is good on the off chance it's called into by some codebase out there. These sort of small PRs also can be quick approve. 

Thanks for the substantial work on the foundations here!

**Comment by VenelinMartinov** [comment]
Thanks for the reviews! After discussing with @t0yv0, I've removed the breaking changes here - we really have no reason to do that. I've cleaned up other unnecessary changes from this PR and will merge once tests pass.
